plugin/pkg/scheduler/schedulercache/nodeinfocache/node_info_cache.go

```diff
@@ -0,0 +1,245 @@
+// "period" is how long the background goroutine should wait before cleaning up expired pods periodically.
+// "stop" is the channel that signals stopping and we would close background goroutines.
+func New(ttl, period time.Duration, stop chan struct{}) schedulercache.NodeInfoCache {
+	cache := newNodeInfoCache(ttl, period, stop)
+	cache.run()
+	return cache
+}
+
+// mustGetPodKey returns the string key of a pod.
+// A pod is ensured to have accessor. We don't want to check the error everytime.
+// TODO: We should consider adding a Key() method to api.Pod
+func mustGetPodKey(pod *api.Pod) string {
+	key, err := clientcache.MetaNamespaceKeyFunc(pod)
+	if err != nil {
+		panic("api.Pod should have key func: " + err.Error())
+	}
+	return key
+}
+
+type nodeInfoCache struct {
+	stop chan struct{}
+
+	// This mutex guards all fields within this cache struct
+	mu          sync.Mutex
+	ttl         time.Duration
+	period      time.Duration
+	assumedPods map[string]assumedPod
+	podStates   map[string]podState
+	// TODO: we should also watch node deletion and clean up node entries
+	nodes map[string]*schedulercache.NodeInfo
+}
+
+type podState int
+
+const (
+	podAsummed podState = iota + 1
+	podAdded
+	podExpired
+)
+
+type assumedPod struct {
+	nodeName string
+	podInfo  schedulercache.PodInfo
+	deadline time.Time
+}
+
+func newNodeInfoCache(ttl, period time.Duration, stop chan struct{}) *nodeInfoCache {
+	return &nodeInfoCache{
+		ttl:    ttl,
+		period: period,
+		stop:   stop,
+
+		nodes:       make(map[string]*schedulercache.NodeInfo),
+		podStates:   make(map[string]podState),
+		assumedPods: make(map[string]assumedPod),
+	}
+}
+
+func (cache *nodeInfoCache) run() {
+	go util.Until(cache.cleanupExpiredAssumedPods, cache.period, cache.stop)
+}
+
+func (cache *nodeInfoCache) GetNodeInfo(nodeName string, cb func(*schedulercache.NodeInfo)) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	n := cache.getNodeInfo(nodeName)
+	cb(n)
+}
+
+// GetNodeInfo returns cached NodeInfo. It returns nil if no such node if found for given node name.
+func (cache *nodeInfoCache) getNodeInfo(nodeName string) *schedulercache.NodeInfo {
+	n, ok := cache.nodes[nodeName]
+	if !ok || n.PodNum == 0 {
+		return nil
+	}
+	return n
+}
+
+func (cache *nodeInfoCache) AssumePod(pod *api.Pod) error {
+	return cache.assumePod(pod, time.Now())
+}
+
+// assumePod exists for making test deterministic by taking time as input argument.
+func (cache *nodeInfoCache) assumePod(pod *api.Pod, now time.Time) error {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	key := mustGetPodKey(pod)
+	if _, ok := cache.podStates[key]; ok {
+		return fmt.Errorf("pod state wasn't initial but get assumed. Pod key: %v", key)
+	}
+
+	cache.addPod(pod)
+	cache.podStates[key] = podAsummed
+	aPod := assumedPod{
+		nodeName: pod.Spec.NodeName,
+		podInfo:  schedulercache.ParsePodInfo(pod),
+		deadline: now.Add(cache.ttl),
+	}
+	cache.assumedPods[key] = aPod
+	return nil
+}
+
```

Comment:
These are internal implementation. We only need to write docs for public API and notes on implicit code. Right?